Modules/Core/Sources/Core/Buffer/Buffer.swift

```diff
@@ -0,0 +1,129 @@
+@_exported import struct Dispatch.DispatchData
+#if os(Linux)
+import Glibc
+#else
+import Darwin
+#endif
+
+public typealias Byte = UInt8
+public typealias Buffer = DispatchData
+
+public protocol BufferInitializable {
+    init(buffer: Buffer) throws
+}
+
+public protocol BufferRepresentable {
+    var buffer: Buffer { get }
+}
+
+extension Buffer : BufferRepresentable {
+    public var buffer: Buffer {
+        return self
+    }
+}
+
+public protocol BufferConvertible : BufferInitializable, BufferRepresentable {}
+
+extension Buffer {
+    public init(_ string: String) {
+        self = [UInt8](string.utf8).withUnsafeBufferPointer { Buffer(bytes: $0) }
+    }
+    
+    public init(_ bytes: [UInt8]) {
+        self = bytes.withUnsafeBufferPointer { Buffer(bytes: $0) }
+    }
+    
+    public init() {
+        self = Buffer.empty
+    }
+    
+    public init(count: Int, fill: (UnsafeMutableBufferPointer<UInt8>) throws -> Void) rethrows {
+        self = try Buffer(capacity: count) {
+            try fill($0)
+            return count
+        }
+    }
+    
+    public init(capacity: Int, fill: (UnsafeMutableBufferPointer<UInt8>) throws -> Int) rethrows {
+        let bytes = UnsafeMutablePointer<UInt8>.allocate(capacity: capacity)
+        let buffer = UnsafeMutableBufferPointer(start: bytes, count: capacity)
+        let usedCapacity = try fill(buffer)
+        
+        guard usedCapacity > 0 else {
+            bytes.deallocate(capacity: capacity)
+            self = Buffer.empty
+            return
+        }
+        
+        guard Double(usedCapacity) > Double(capacity) * 0.25 else {
+            defer {
+                bytes.deallocate(capacity: capacity)
+            }
+            self = Buffer(bytes: UnsafeBufferPointer<UInt8>(start: bytes, count: usedCapacity))
+            return
+        }
+        
+        self = Buffer(bytesNoCopy: UnsafeBufferPointer<UInt8>(start: bytes, count: usedCapacity), deallocator: .free)
+    }
+    
+    public subscript(_ range: Range<Int>) -> Buffer {
+        return subdata(in: self.startIndex.advanced(by: range.lowerBound)..<self.startIndex.advanced(by: range.upperBound))
```

Comment:
I don't think we should -- it's convenience over the start/end index which is going to throw fatal errors anyways.